src/coord/src/coord.rs

```diff
@@ -2040,13 +2085,68 @@ impl Coordinator {
 
     async fn sequence_peek(
         &mut self,
-        conn_id: u32,
+        session: &mut Session,
         source: MirRelationExpr,
         when: PeekWhen,
         finishing: RowSetFinishing,
         copy_to: Option<CopyFormat>,
     ) -> Result<ExecuteResponse, CoordError> {
-        let timestamp = self.determine_timestamp(&source, when)?;
+        let conn_id = session.conn_id();
+        let in_transaction = matches!(
+            session.transaction(),
+            &TransactionStatus::InTransaction(_) | &TransactionStatus::InTransactionImplicit(_)
+        );
+        // For explicit or implicit transactions that do not use AS OF, get the
+        // timestamp of the in-progress transaction or create one. If this is an AS OF
+        // query, we don't care about any possible transaction timestamp. If this is a
+        // single-statement transaction (TransactionStatus::Started), we don't need to
+        // worry about preventing compaction or choosing a valid timestamp for future
+        // queries.
+        let timestamp = if in_transaction && when == PeekWhen::Immediately {
+            let timestamp = session.get_transaction_timestamp(|| {
+                // Determine a timestamp that will be valid for anything in any schema
+                // referenced by the first query. This is a first pass implementation of "time
+                // domains".
+                let ids = self.catalog.timedomain_for(&source, conn_id);
+
+                // We want to prevent compaction of the indexes consulted by
+                // determine_timestamp, not the ones listed in the query.
+                let (timestamp, index_ids) =
+                    self.determine_timestamp(&ids, PeekWhen::Immediately)?;
+                let mut ids = HashSet::new();
+                let mut handles = vec![];
+                for id in index_ids {
+                    handles.push(self.indexes.get(&id).unwrap().since_handle(vec![timestamp]));
+                    ids.insert(id);
+                }
+                self.txn_reads.insert(
+                    conn_id,
+                    TxnReads {
+                        ids,
+                        _handles: handles,
+                    },
+                );
+
+                Ok(timestamp)
+            })?;
+
+            // Verify that the indexes for this query are in the current read transaction.
+            // TODO: we should probably compare the results of timedomain_for instead of nearest_indexes.
+            let (index_ids, _indexes_complete) =
```

Comment:
Collecting the indexes only unfortunately leaves non-materialized views without transactional protection:

```
> CREATE VIEW v1 AS SELECT 1;

> BEGIN

> SELECT * FROM v1;
1

$ postgres-execute connection=postgres://materialize:materialize@${testdrive.materialized-addr}
CREATE OR REPLACE VIEW v1 AS SELECT 2;
SELECT * FROM v1 AS OF NOW();

> SELECT * FROM v1;
1
```
The last query returns 2 even though it is in the same transaction as the query that perviously returned 1. This should have either returned "1" (which I guess is tough to accompilish) or return an error.

I can open a separate ticket for this if you think this should not / can not be addressed in the current PR.

Reply:
Fixed.